Deprecate prototype usage

benjaminwood opened the pull request:
This PR deprecates all public methods for several prototype related modules.

Deprecating usage will help us to safely remove all traces from the app.

benjaminwood commented:
I've added a condition checking the env var `DEPRECATE_PROTOTYPE`. This allows us to disable the deprecation under certain circumstances (it's real noisy right now). If `DEPRECATE_PROTOTYPE` is equal to `'false'` (string), the deprecation will be disabled.

I also did a bit of analysis on the deprecation warnings generated by CI. Here's a summary:

```
COUNT    | LINE
-----------------------------------------------------
      49 | javascript_tag
       8 | link_to_remote
       8 | remote_form_for
       7 | escape_javascript
       7 | remote_function
       5 | sortable_element
       4 | link_to_function
       3 | update_page
       2 | button_to_remote
       1 | button_to_function
       1 | form_remote_tag
       1 | javascript_cdata_section
       1 | observe_form
       1 | sortable_element_js
       1 | visual_effect
```

And a deduplicated list of deprecated methods with locations:

```
Job #1

#link_to_remote /rigse/app/helpers/application_helper.rb:679
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:178
#escape_javascript /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/prototype_helper.rb:134
#link_to_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:178
#remote_function /rigse/app/helpers/application_helper.rb:203
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_cdata_section /rigse/config/initializers/00_deprecate_prototype.rb:12
#javascript_tag /rigse/app/views/layouts/application.html.haml:55
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#escape_javascript /rigse/app/views/shared/_footer.html.haml:17
#javascript_tag /rigse/app/views/layouts/application.html.haml:53
#visual_effect /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/scriptaculous_helper.rb:237
#link_to_function /rigse/app/helpers/application_helper.rb:615
#escape_javascript /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/prototype_helper.rb:140
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#javascript_tag /rigse/app/views/layouts/application.html.haml:53
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:51
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:66
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:51
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:66
#remote_form_for /rigse/app/views/portal/clazzes/_remote_form.html.haml:1
#form_remote_tag /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:279
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:220
#button_to_remote /rigse/app/helpers/portal/teachers_helper.rb:32
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:7
#button_to_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:7
#remote_form_for /rigse/app/views/portal/clazzes/_remote_form.html.haml:1
#sortable_element /rigse/app/views/shared/_offering_for_teacher.html.haml:26
#sortable_element_js /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/scriptaculous_helper.rb:133
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/scriptaculous_helper.rb:138
#javascript_tag /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/scriptaculous_helper.rb:133
#link_to_remote /rigse/app/views/portal/students/_table_for_clazz.html.haml:33
#update_page /rigse/app/views/portal/schools/destroy.js.rjs:1
#remote_form_for /rigse/app/views/portal/students/_register.html.haml:1
#link_to_function /rigse/app/helpers/application_helper.rb:622
#update_page /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/javascript_helper.rb:59
#link_to_function /rigse/app/helpers/application_helper.rb:688
#remote_form_for /rigse/app/views/embeddable/open_responses/_remote_form.html.haml:1
#javascript_tag /rigse/app/views/external_activities/_form.html.haml:121
#observe_form /rigse/app/views/images/index.html.haml:3
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:424
#javascript_tag /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:430
#sortable_element /rigse/app/views/materials_collections/_materials_in_collection.html.haml:23

Job #2

#remote_form_for /rigse/app/views/portal/students/_register.html.haml:1
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#escape_javascript /rigse/app/views/shared/_footer.html.haml:17
#javascript_tag /rigse/app/views/layouts/application.html.haml:55
#escape_javascript /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/action_view/helpers/prototype_helper.rb:634
#javascript_tag /rigse/app/views/layouts/application.html.haml:53
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#javascript_tag /rigse/app/views/interactives/_form.html.haml:47
#sortable_element /rigse/app/views/materials_collections/_materials_in_collection.html.haml:23
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:51
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:53
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:57
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:66
#sortable_element /rigse/app/views/shared/_offering_for_teacher.html.haml:26
#link_to_remote /rigse/app/views/home/_notice.html.haml:11
#link_to_remote /rigse/app/views/home/_notice.html.haml:38

Job #3

#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#escape_javascript /rigse/app/views/shared/_footer.html.haml:17
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#javascript_tag /rigse/app/views/layouts/application.html.haml:55
#link_to_remote /rigse/app/views/home/_notice.html.haml:11
#link_to_remote /rigse/app/views/home/_notice.html.haml:38
#javascript_tag /rigse/app/views/layouts/application.html.haml:53
#remote_form_for /rigse/app/views/admin/settings/_remote_form.html.haml:1
#remote_function /rigse/vendor/bundle/ruby/2.2.0/bundler/gems/prototype_legacy_helper-39efacca4fd1/lib/prototype-rails/prototype_legacy_helper.rb:220)
#remote_form_for /rigse/app/views/portal/students/_register.html.haml:1
#javascript_tag /rigse/app/views/interactives/_form.html.haml:47
#sortable_element /rigse/app/views/materials_collections/_materials_in_collection.html.haml:23
#remote_form_for /rigse/app/views/materials_collections/_remote_form.html.haml:1
#button_to_remote /rigse/app/helpers/portal/student_clazzes_helper.rb:42
#link_to_remote /rigse/app/views/portal/students/_table_for_clazz.html.haml:33
#update_page /rigse/app/views/portal/student_clazzes/destroy.js.rjs:1
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:51
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:53
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:57
#javascript_tag /rigse/app/views/layouts/minimal.html.haml:66

Job #4

#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
#escape_javascript /rigse/app/views/shared/_footer.html.haml:17
#javascript_tag /rigse/app/views/external_activities/_form.html.haml:121
#javascript_tag /rigse/app/views/layouts/application.html.haml:55
#javascript_tag /rigse/app/views/layouts/application.html.haml:53
#link_to_remote /rigse/app/views/search/_material_unassigned_clazzes.html.haml:95
#javascript_tag /rigse/app/views/layouts/application.html.haml:49
#javascript_tag /rigse/app/views/layouts/application.html.haml:51
#javascript_tag /rigse/app/views/layouts/application.html.haml:64
```

benjaminwood commented:
Note some methods like `javascript_tag` are replaced in later versions of rails with compatible methods. So, I do not believe any action will be necessary on those.

scytacki commented:
@benjaminwood is this PR still something to merge?
It is failing on travis: https://travis-ci.org/concord-consortium/rigse/jobs/424567948#L983

benjaminwood commented:
Thanks @scytacki I added an additional namespace in my last commit and didn't realize it broke the build. I've added a fix for that and yes I think this should still be merged as it'll help us ensure prototype usage is removed.

scytacki commented:
I don't understand why, but github is not reporting the failing build in travis for this branch: https://travis-ci.org/concord-consortium/rigse/builds/426812984
It looks un-related though.
